File: VB/ConsoleClient/ClientProgram.vb
Imports DevExpress.ExpressApp.Security.ClientServer.Wcf
Imports System.ServiceModel
Imports DevExpress.ExpressApp.Security.ClientServer
Imports DevExpress.ExpressApp.Security
Imports DevExpress.Xpo.Metadata
Imports System.Reflection
Imports PersistentClassesLibrary
Imports System
Imports DevExpress.Xpo
Imports System.Collections.Generic
Imports DevExpress.Data.Filtering
Imports System.Configuration
Namespace ConsoleClient
    Friend Class Program

       Private Shared Sub AccessEmployee(ByVal clientDataServer As WcfSecuredDataServerClient, ByVal statelessClientInfo As ClientInfo)
            Try
                Dim dictionary As XPDictionary = New ReflectionDictionary()
                dictionary.CollectClassInfos(System.Reflection.Assembly.GetAssembly(GetType(Employee)))
                Dim isReadGranted As Boolean = clientDataServer.IsGranted(statelessClientInfo, New SerializablePermissionRequest(GetType(Employee), Nothing, Nothing, SecurityOperations.Read))
                Console.WriteLine(String.Format("Is read granted for Employee: {0}", isReadGranted))

                Dim isWriteGranted As Boolean = clientDataServer.IsGranted(statelessClientInfo, New SerializablePermissionRequest(GetType(Employee), Nothing, Nothing, SecurityOperations.Write))
                Console.WriteLine(String.Format("Is modification granted for Employee: {0}", isWriteGranted))

                Dim securedObjectLayerClient As New SecuredSerializableObjectLayerClient(statelessClientInfo, clientDataServer)
                Dim client As New SerializableObjectLayerClient(securedObjectLayerClient, dictionary)
                Dim uow As New UnitOfWork(client)
                Console.WriteLine(ControlChars.Lf & "Read all available Employee objects:")
                For Each obj As Employee In New XPCollection(uow, GetType(Employee))
                    Console.WriteLine(obj.Name)
                Next obj
                Console.WriteLine(ControlChars.Lf & "Try to modify object: ")
                Try
                    Dim employee As Employee = uow.FindObject(Of Employee)(New BinaryOperator("Name", "Peter"))
                    Console.WriteLine("Object name: " & employee.Name)
                    employee.Name &= "_modified"
                    uow.CommitChanges()
                    Console.WriteLine("New object name: " & employee.Name)
                Catch e As Exception
                    Console.WriteLine("Error occured: " & e.Message)
                End Try
            Catch e As Exception
                Console.WriteLine(e.Message & Environment.NewLine & e.StackTrace)
            End Try
            Console.WriteLine()
       End Sub

        Shared Sub Main(ByVal args() As String)
            Dim clientDataServer As New WcfSecuredDataServerClient(WcfDataServerHelper.CreateDefaultBinding(), New EndpointAddress("http://localhost:1424/DataServer"))

            Console.WriteLine("Login as 'User' user")
            Dim statelessClientInfo1 As New ClientInfo(Guid.Empty, Guid.Empty, New AuthenticationStandardLogonParameters("User", ""))
            AccessEmployee(clientDataServer, statelessClientInfo1)

            Console.WriteLine(ControlChars.Lf & "Login as 'Admin' user")
            Dim statelessClientInfo2 As New ClientInfo(Guid.Empty, Guid.Empty, New AuthenticationStandardLogonParameters("Admin", ""))
            AccessEmployee(clientDataServer, statelessClientInfo2)

            Console.WriteLine(ControlChars.Lf & "Press Enter to close...")
            Console.ReadLine()
        End Sub
    End Class
End Namespace

File: CS/ConsoleClient/ClientProgram.cs
using DevExpress.ExpressApp.Security.ClientServer.Wcf;
using System.ServiceModel;
using DevExpress.ExpressApp.Security.ClientServer;
using DevExpress.ExpressApp.Security;
using DevExpress.Xpo.Metadata;
using System.Reflection;
using PersistentClassesLibrary;
using System;
using DevExpress.Xpo;
using System.Collections.Generic;
using DevExpress.Data.Filtering;
using System.Configuration;
namespace ConsoleClient {
    class Program {
        const string connectionString = "http://localhost:1424/DataServer";
        static XPDictionary dictionary = new ReflectionDictionary();
        
        static void Main(string[] args) {
            try {
                XPDictionary dictionary = new ReflectionDictionary();
                dictionary.CollectClassInfos(Assembly.GetAssembly(typeof(Employee)));

                Console.WriteLine("Login as 'User' user");
                ClientInfo statelessClientInfo = new ClientInfo(Guid.Empty, Guid.Empty, new AuthenticationStandardLogonParameters("User", ""));
                WcfSecuredDataServerClient clientDataServer = new WcfSecuredDataServerClient(
                    WcfDataServerHelper.CreateDefaultBinding(), new EndpointAddress("http://localhost:1424/DataServer"));

                bool isReadGranted = clientDataServer.IsGranted(statelessClientInfo, new ClientPermissionRequest(typeof(Employee), null, null, SecurityOperations.Read));
                Console.WriteLine(string.Format("Is read granted for Employee: {0}", isReadGranted));

                bool isWriteGranted = clientDataServer.IsGranted(statelessClientInfo, new ClientPermissionRequest(typeof(Employee), null, null, SecurityOperations.Write));
                Console.WriteLine(string.Format("Is modification granted for Employee: {0}", isWriteGranted));

                SecuredSerializableObjectLayerClient securedObjectLayerClient = new SecuredSerializableObjectLayerClient(statelessClientInfo, clientDataServer);
                SerializableObjectLayerClient client = new SerializableObjectLayerClient(securedObjectLayerClient, dictionary);
                UnitOfWork uow = new UnitOfWork(client);
                Console.WriteLine("Read all available Employee objects:");
                foreach (Employee obj in new XPCollection(uow, typeof(Employee))) {
                    Console.WriteLine(obj.Name);
                }
                Console.WriteLine("Try to modify object: ");
                try {
                    Employee employee = uow.FindObject<Employee>(new BinaryOperator("Name", "Available object"));
                    Console.WriteLine("Object name: " + employee.Name);
                    employee.Name += "_modified";
                    uow.CommitChanges();
                }
                catch(Exception e) {
                    Console.WriteLine("Error occured: " + e.Message);
                }
            } catch (Exception e) {
                Console.WriteLine(e.Message + Environment.NewLine + e.StackTrace);
            }
            Console.WriteLine("Press Enter to close...");
            Console.ReadLine();
        }
    }
}
